### PR TITLE
Refactor: Move transaction testing utilities from `crates/chain/tests/common` to `testenv` crate

### LINKED ISSUE #1602: Move stuff from `crates/chain/tests/common` to `bdk_testenv` and make them publicly-accessible
**Describe the enhancement**  

These macros, fake-tx-construction-tools are helpful in other crates too. Let's move them to `bdk_testenv` and make them publicly available. 

**Use case**  

Better test tooling.

### PR DESCRIPTION
**Pull request by tvpeter**
### Description

This change moves all transaction testing utilities (macros and functions) from `crates/chain/tests/common` to `crates/testenv/tx_utils` to minimize duplication.

Closes #1602 

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

### CODE REVIEW

**Comment by evanlinjin** [comment]
Hey @tvpeter thank you for taking this on! Please follow the [https://www.conventionalcommits.org/en/v1.0.0/](Conventional Commit Specification) for the commit messages. I'll review this properly soon!

**Comment by tvpeter** [comment]
@evanlinjin thank you.
I have reworded the commit messages to follow the specification.

File: crates/testenv/Cargo.toml [hunk 1]
```diff
@@ -18,6 +18,7 @@ workspace = true
 [dependencies]
 bdk_chain = { path = "../chain", version = "0.19.0", default-features = false }
 electrsd = { version = "0.28.0", features = [ "bitcoind_25_0", "esplora_a33e97e1", "legacy" ] }
+bitcoin = { version = "0.32.0", default-features = false }
```
Comment (LagginTimes) [inline crates/testenv/Cargo.toml]: Hello! Since `bdk_chain` already re-exports `bitcoin`, we don't need to add `bitcoin` as a dependency here.

Comment (tvpeter) [reply crates/testenv/Cargo.toml]: Fixed. Thank you

**Review by oleonardolima — COMMENTED** [review]
Concept ACK

I only wonder about naming/structure, maybe reword it to only `utils.rs` or have both `chain_utils.rs` and `tx_utils.rs`, but I'm not sure about it.

**Comment by tvpeter** [comment]
Alright @oleonardolima. I initially thought that it will contain transaction related utilities only but on a second thought most of the utilities are not just for transactions. So I will rename it to just `utils.rs` and move the `descriptors` sample array there too.  Thank you.

**Comment by evanlinjin** [comment]
Please use the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) format for commit messages.

Looking at this now, maybe we should create a `bdk_testutil` crate for this instead? These are light macros, functions. Whereas `bdk_testenv` requires downloading binaries during the build process.

cc. @ValuedMammal @notmandatory

**Comment by ValuedMammal** [comment]
> Looking at this now, maybe we should create a `bdk_testutil` crate for this instead? These are light macros, functions. Whereas `bdk_testenv` requires downloading binaries during the build process.

Another option that comes to mind is to have conditional compilation of testenv with something like a "full" feature flag, but I don't have a strong opinion on it at the moment.

**Comment by LagginTimes** [comment]
> > Looking at this now, maybe we should create a `bdk_testutil` crate for this instead? These are light macros, functions. Whereas `bdk_testenv` requires downloading binaries during the build process.
> 
> Another option that comes to mind is to have conditional compilation of testenv with something like a "full" feature flag, but I don't have a strong opinion on it at the moment.

I think conditional compilation sounds like the better option here vs. having an additional crate to manage, since both crates would be primarily focused on testing.

**Comment by ValuedMammal** [comment]
> > > Looking at this now, maybe we should create a `bdk_testutil` crate for this instead? These are light macros, functions. Whereas `bdk_testenv` requires downloading binaries during the build process.
> > 
> > 
> > Another option that comes to mind is to have conditional compilation of testenv with something like a "full" feature flag, but I don't have a strong opinion on it at the moment.
> 
> I think conditional compilation sounds like the better option here vs. having an additional crate to manage, since both crates would be primarily focused on testing.

~~Considering the test utils are common to other crates and change infrequently, it might be just as easy to move them to a `test_util` module in `bdk_core`~~

Edit: As discussed on dev call the current direction on the PR makes sense

**Comment by LagginTimes** [comment]
Nit: I think squashing the commits down to one would make the history a bit cleaner.

**Comment by tvpeter** [comment]
> Nit: I think squashing the commits down to one would make the history a bit cleaner.

Done

File: crates/chain/tests/test_local_chain.rs [hunk 2]
```diff
@@ -9,6 +9,7 @@ use bdk_chain::{
     },
     BlockId,
 };
+use bdk_testenv::{chain_update, h, local_chain};
```
Comment (oleonardolima) [inline crates/chain/tests/test_local_chain.rs]: As we are importing this now from bdk_testenv, and deleted the macros from `commond/mod.rs` I guess we can remove the code at L16-17.

``` rust
#[macro_use]
mod common;
```

File: crates/testenv/src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    }};
+}
+
+#[allow(unused_macros)]
+#[macro_export]
+macro_rules! h {
+    ($index:literal) => {{
+        bitcoin::hashes::Hash::hash($index.as_bytes())
+    }};
+}
```
Comment (oleonardolima) [inline crates/testenv/src/utils.rs]: IIRC we discussed during the call that this could have a meaningful name, I  would use `hash` for instance.

**Review by oleonardolima — COMMENTED** [review]
I did another review and left some comments. Overall, it's looking good!

I think you still need to change the commit message to lowercase.

**Comment by oleonardolima** [comment]
You also probably need to rebase on top of the master, so the CI issues are fixed (in case you are curious, check: #1636).

**Comment by tvpeter** [comment]
> I did another review and left some comments. Overall, it's looking good!
> 
> I think you still need to change the commit message to lowercase.

Done. Thank you

File: crates/testenv/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+#[allow(unused_macros)]
+#[macro_export]
+macro_rules! local_chain {
+    [ $(($height:expr, $block_hash:expr)), * ] => {{
+        #[allow(unused_mut)]
+        bdk_chain::local_chain::LocalChain::from_blocks([$(($height, $block_hash).into()),*].into_iter().collect())
+            .expect("chain must have genesis block")
+    }};
+}
```
Comment (LagginTimes) [inline crates/testenv/src/utils.rs]: Nit: I suggest renaming this expression to `hash` for consistency with the other macros.

Comment (tvpeter) [reply crates/testenv/src/utils.rs]: The previous `h!` macro has been renamed to `hash!`, so it will conflict. Since it builds a `LocalChain`, is it better to rename it `new_local_chain`?

Comment (LagginTimes) [reply crates/testenv/src/utils.rs]: Oops! Just to clarify: I meant to change `block_hash` to `hash`, not rename the `local_chain` macro. Apologies for the confusion!

Comment (tvpeter) [reply crates/testenv/src/utils.rs]: done

**Review by ValuedMammal — APPROVED** [review]
ACK 88a8403707038264251b3e7fa18b8f21af8646ff

**Review by notmandatory — APPROVED** [review]
ACK 88a8403707038264251b3e7fa18b8f21af8646ff

Thanks for taking this one on.
